design/ingressroute-design.md

```diff
@@ -0,0 +1,398 @@
+kind: IngressRoute
+metadata:
+  name: gmail 
+  namespace: gmail
+spec:
+  routes:
+  - match: /mail
+    service:
+      name: gmail-tomcat-svc
+      port: 8000
+```
+That's all fine, the operators in the `gmail` namespace can't change the delegation in the `virtualhost-namespace` and the admins in the `virtualhost-namespace` feel confident that they control `www.google.com` and that gmail is being served over tls.
+
+However, at some time later ad money is getting tight and the operators take on a new customer, who is very cheap and doesn't want to pay for tls.
+```yaml
+apiVersion: contour.heptio.com/v1alpha1
+kind: IngressRoute
+metadata:
+  name: daves-cheap-webhosting
+  namespace: virtualhost-namespace
+spec:
+  virtualhost:
+    fqdn: dave.cheney.net
+  routes:
+  - match: /
+    delegate:
+      name: webapp
+      namespace: customer-dave
+```
+Because the ingress operators don't want to have to handle tickets for dave's cheap webhosting, they delegate `/` to dave's namespace and tell him if he has a problem getting it working he can ask for help on the forum.
+
+_However_, dave deploys this vertex ingressroute CRD.
+```yaml
+apiVersion: contour.heptio.com/v1alpha1
+kind: IngressRoute
+metadata:
+  name: webapp 
+  namespace: customer-dave
+spec:
+  routes:
+  - match: /mail
+    delegate:
+      name: gmail
+      namespace: gmail
+```
+And all of a sudden, `http://dave.cheney.net/mail` is serving up the full gmail application, sans TLS.
+
+This is a large worked example, but the problem boils down to "vertices have no way to list whom they are delegated too".
+Without a way to deny or permit delegations, anyone who knows the name of the object (which I don't believe is a secret, and even if it was, is security via obscurity) can expose those routes on a virtualhost they own.
+
+### Mitigations
+
+The best way I've thought to mitigate this so far is the vertex ingress routes would have to list the name of either
+
+- the name/namespace of the incoming delegation
+- the virtualhost.fqdn at the root of a delegation chain.
+
+This is unfortunate for two reasons,
+
+1. It's more boilerplate; a deletates to b, b has to list a as a cross check.
+2. It turns a DAG into a linked list (possibly not the right term), but it would be impossible for a web service to be a member of multiple roots, unless of course we made the list of incoming vertices be a list -- which would probably push the solution into using virtualhost.fqdn.
+
+The last thing is that this boilerplate _should be required_ even when not in enforcing mode.
+I'm not interested in proposing a design where the security interlock that prevents dave's cheap webhosting from publishing gmail without TLS is considered to be the customers problem because they did not add an optional key. 
+Said again, if this is the mitigation we choose to adopt, it has to be mandatory for all users, because we all know how effective optional security features are.
```

Comment:
I'm not convinced that this is a huge security issue.  The attacker could proxy to the gmail service via any number of methods.

The thing that I think is more concerning is crossing the linkages between different ingress "worlds".  Imagine that I have one contour setup for external exposure to the network and another for internal exposure of dashboards and stuff (perhaps via VPN, corp network or behind an OAuth proxy).  If we don't have a specification of which "world" an IngressRoute is in than it is easy to cross talk on accident.  Having a an "ingress scope" that gets carried down the DAG might help to mitigate that.  I'd like to explore that (along with making the ingress route be completely absolute including vhost).

Reply:
The notion of scope that travels downwards is interesting, but I think it's going to be hard to reason about completely. Even if scope is applied at the root level (rather than being a property of a path traversal) then, unless you're in enforcing mode with root objects restricted, there is nothing to prevent me creating my own root and ignoring the scope applied from your root.

I think part of the problem is the security model of "enforcing mode" have not been well defined, so its just a bunch of ideas which I thought were a good idea.